shell/shared/grain.js

```diff
@@ -386,6 +389,78 @@ Meteor.methods({
       return outerResult;
     }
   },
+  requestAccess: function (origin, grainId, identityId) {
+    check(origin, String);
+    check(grainId, String);
+    check(identityId, String);
+    if (!this.isSimulation) {
+      if (!this.userId) {
+        throw new Meteor.Error(403, "Must be logged in to request access.");
+      }
+      if (!globalDb.userHasIdentity(this.userId, identityId)) {
+        throw new Meteor.Error(403, "Not an identity of the current user: " + identityId);
+      }
+
+      const grain = Grains.findOne(grainId);
+      if (!grain) {
+        throw new Meteor.Error(404, "No such grain");
+      }
+      if (grain.accessRequests) {
+        const previous = _.findWhere(grain.accessRequests, {identityId: identityId});
+        if (previous) {
+          throw new Meteor.Error(403, "Not allowed to request access more than once. " +
+                                 "Previous request was sent on " + previous.sent);
+        }
+      }
+      const grainOwner = globalDb.getUser(grain.userId);
+      const email = _.findWhere(SandstormDb.getUserEmails(grainOwner), {primary: true});
+      if (!email) {
+        throw new Meteor.Error("no email", "Grain owner has no email address.");
+      }
+      const emailAddress = email.email;
+
+      const identity = globalDb.getIdentity(identityId);
+      globalDb.addContact(grainOwner._id, identityId);
+
+      const envelopeFrom = globalDb.getReturnAddress();
+      let fromEmail = globalDb.getServerTitle() + " <" + globalDb.getReturnAddress() + ">";
+      const senderEmails = SandstormDb.getVerifiedEmails(identity);
+      if (senderEmails.length > 0) {
+        fromEmail = senderEmails[0];
+      }
+
+      // TODO(soon): In the HTML version, we should display an identity card.
+      let identityNote = "";
+      if (identity.profile.service === "google") {
+        identityNote = " (" + identity.privateIntrinsicName + ")";
+      } else if (identity.profile.service === "github") {
+        identityNote = " (" + identity.profile.intrinsicName + " on GitHub)";
+      } else if (identity.profile.service === "email") {
+        identityNote = " (" + identity.profile.intrinsicName + ")";
+      }
+
+      const message = identity.profile.name + identityNote +
+            " has requested access to your grain: " + grain.title + ".";
```

Comment:
Prefer present tense here: "is requesting access" rather than "has requested access".